src/internal/connector/exchange/service_restore.go

```diff
@@ -651,19 +651,27 @@ func establishContactsRestoreLocation(
 
 	ctx = clues.Add(ctx, "is_new_cache", isNewCache)
 
-	temp, err := ac.Contacts().CreateContactFolder(ctx, user, folders[0])
-	if err != nil {
+	fold, err := ac.Contacts().CreateContactFolder(ctx, user, folders[0])
+	// 409 handling: Fetch folder if it exists.
+	// This is rare, but it may happen if an earlier CreateContactFolder POST failed
+	// with 5xx, potentially leaving dirty state in graph.
+	if graph.IsErrFolderExists(err) {
+		fold, err = ac.Contacts().GetContactFolderByDisplayName(ctx, user, folders[0])
+		if err != nil {
+			return "", err
+		}
+	} else if err != nil {
```

Comment:
Very minor: I prefer the following code ordering.  It establishes the worst case/immediate exits at the top, and gradually moves towards happy path handling. 
```suggestion
	if err != nil && !graph.IsErrFolderExists(err) {
	  return "", err
	}
	
	if graph.IsErrFolderExists(err)
		fold, err = ac.Contacts().GetContactFolderByDisplayName(ctx, user, folders[0])
		if err != nil {
			return "", err
		}
	}
```

Reply:
Kinda subjective.  I didn't want to call `IsErrFolderExists` twice. But your pattern seems better( assuming we use it throughout the codebase). I'll make this change.